Adding ExecutionSpace partitioning function

crtrott opened the pull request:
This allows the creation of multiple instances, but only CUDA implements an actual implementation, everyone else returns the same thing multiple times for now.

The design is as a customization point, which would allow for ADL hopefully.

DavidPoliakoff commented on core/src/Kokkos_Core.hpp:
```diff
@@ -274,6 +274,23 @@ class ScopeGuard {
+//   Customization point for backends
+//   Default behavior is to return the passed in instance
+template <class ExecSpace, class... Args>
+std::array<ExecSpace, sizeof...(Args)> partition_space(ExecSpace space,
```
Are we intentionally requiring people to partition into a compile-time sized set of args? Should we have a variant where this takes a vector of weights and returns a vector of instances?

crtrott replied:
Hm interesting question, we could have both, that said I didn't see a real need for that much runtime choice.

PhilMiller replied:
The partition count in EMPIRE will be runtime-determined

PhilMiller replied:
That said, it could just call the current unweighted CUDA implementation that just makes separate streams one by one, but that's probably not the usage you were going for.

crtrott replied:
done, using std::vector now

PhilMiller commented on core/src/Kokkos_Cuda.hpp:
```diff
@@ -251,6 +251,21 @@ class Cuda {
+  std::array<Cuda, sizeof...(Args)> instances;
+  for (int s = 0; s < int(sizeof...(Args)); s++) {
+    cudaStream_t stream;
+    cudaStreamCreate(&stream);
```
This is missing an error check

crtrott replied:
done

PhilMiller commented on core/src/Cuda/Kokkos_Cuda_Instance.cpp:
```diff
@@ -711,6 +713,8 @@ void CudaInternal::finalize() {
         Kokkos::kokkos_free<Kokkos::CudaSpace>(m_team_scratch_ptr[i]);
     }
 
+    if (m_manage_stream && m_stream != nullptr) cudaStreamDestroy(m_stream);
```
Missing error check here, too.

PhilMiller commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -0,0 +1,85 @@
+void check_equalness(ExecSpace, ExecSpace) {}
+
+#ifdef KOKKOS_ENABLE_CUDA
+void check_equalness(Kokkos::Cuda exec1, Kokkos::Cuda exec2) {
```
Perhaps `check_distinctive`?

crtrott replied:
done

PhilMiller commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -0,0 +1,85 @@
+      sum1);
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<TEST_EXECSPACE>(instances[1], 0, N), SumFunctor(),
+      sum2);
```
This poses the same testing challenge as I noted in #4059 - confirming that various kernels actually ran on distinct execution spaces, and that they were at least hypothetically concurrent.

crtrott replied:
the distinctive part is probably good enough for now

crtrott commented:
- Allow runtime size via some vector type. 
- Return always std::vector.
- No normalization requirement.
- Support float and int, but make sure its either, not a mix?

crtrott commented:
I addressed all the points: for the int/float thing I allow mix, I am using a fold expression to static_assert that they all are, that fold expresssion is protected by feature test macro.

crtrott commented:
@PhilMiller @DavidPoliakoff I think i addressed everything

PhilMiller commented:
For runtime-determined partition count, I think the partition functions need to take `std::vector` rather than variadic arguments, too?

DavidPoliakoff commented:
@PhilMiller yeah, I believe we discussed this in the meeting. Should either be a vector, or a template<template<typename> typename> thingy, and I recommend vector

masterleinad commented on core/src/Kokkos_Cuda.hpp:
```diff
@@ -251,6 +251,28 @@ class Cuda {
+// relative weight
+//   Customization point for backends
+//   Default behavior is to return the passed in instance
+template <class... Args>
+std::vector<Cuda> partition_space(Cuda space, Args...) {
+  std::vector<Cuda> instances(sizeof...(Args));
+#ifdef __cpp_fold_expressions
+  static_assert(
+      (... && std::is_arithmetic_v<Args>),
+      "Kokkos Error: partitioning arguments must be integers or floats");
+#endif
+  for (int s = 0; s < int(sizeof...(Args)); s++) {
+    cudaStream_t stream;
+    CUDA_SAFE_CALL(cudaStreamCreate(&stream));
+    instances[s] = Cuda(stream, true);
+  }
+  return instances;
+}
```
```
/var/jenkins/workspace/Kokkos/core/src/Kokkos_Cuda.hpp:260:40: error: unused parameter 'space' [clang-diagnostic-unused-parameter]
std::vector<Cuda> partition_space(Cuda space, Args...) {
                                       ^
/var/jenkins/workspace/Kokkos/core/src/Kokkos_Cuda.hpp:269:5: error: use of undeclared identifier 'CUDA_SAFE_CALL' [clang-diagnostic-error]
    CUDA_SAFE_CALL(cudaStreamCreate(&stream));
    ^
```

PhilMiller commented:
I think the variadic version only needs to be implemented once, in the generic code, since it can then call the version taking a `vector<T> weights` and let the overload resolution happen there.

PhilMiller reviewed (approved):
This will suite EMPIRE's needs well enough. The interpretation of the weights will become an interesting question down the line, but it'll be good to get this step integrated

DavidPoliakoff reviewed (approved):
I'm very lightly opposed to having a vector and a variadic overload. If anybody else wants to back this up, I'll change it to a request changes. But I think this is a good implementation of what we're aiming at

dalg24 commented on core/src/HIP/Kokkos_HIP_Instance.hpp:
```diff
@@ -159,6 +162,45 @@ class HIPInternal {
+}
+
+template <class T>
+std::vector<HIP> partition_space(const HIP &, std::vector<T> &weights) {
```
Why are `weights` by non-const reference?

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -0,0 +1,117 @@
+};
+
+template <class ExecSpace>
+void check_distinctive(ExecSpace, ExecSpace) {}
```
You mean distinct from one another?

dalg24 commented on core/unit_test/TestExecSpacePartitioning.hpp:
```diff
@@ -0,0 +1,117 @@
+
+TEST(TEST_CATEGORY, partitioning_by_args) {
+  auto instances =
+      Kokkos::Experimental::partition_space(TEST_EXECSPACE(), 1, 1.);
```
Did we capture anywhere the rational for being able to mix types?
